cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,231 @@
+	"github.com/projectcalico/libcalico-go/lib/backend/k8s"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/ipam"
+	"github.com/projectcalico/libcalico-go/lib/logutils"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/options"
+)
+
+const (
+	ipAllocPath = "/var/lib/cni/networks/k8s-pod-network"
+)
+
+func main() {
+	// nodename associates IPs to this node.
+	nodename := os.Getenv("KUBERNETES_NODE_NAME")
+	logCtxt := log.WithField("node", nodename)
+
+	if msg, err := Main(nodename); err != nil {
+		logCtxt.WithError(err).Fatal(msg)
+	}
+}
+
+func Main(nodename string) (msg string, err error) {
+	// Declare variables to avoid shadowing err.
+	var (
+		cfg          *apiconfig.CalicoAPIConfig
+		calicoClient client.Interface
+		clusterInfo  *v3.ClusterInformation
+		node         *v3.Node
+		files        []os.FileInfo
+		ip           *cnet.IP
+		b            []byte
+		calicoDS     *v1beta1.DaemonSet
+		podList      *corev1.PodList
+	)
+
+	// Set up logging formatting.
+	log.SetFormatter(&logutils.Formatter{})
+
+	// Install a hook that adds file/line no information.
+	log.AddHook(&logutils.ContextHook{})
+
+	// Check to see if the system is still using host-local
+	// by checking the existence of the path.
+	if _, err = os.Stat(ipAllocPath); err != nil && os.IsNotExist(err) {
+		log.Info("no migration necessary, successfully exiting...")
+		return
+	}
+
+	//
+	// Initialize necessary variables.
+
+	// ctxt for network calls
+	ctxt := context.Background()
+
+	// calicoClient makes IPAM calls.
+	cfg, err = apiconfig.LoadClientConfig("")
+	if err != nil {
+		msg = "failed to load api client config"
+		return
+	}
+	cfg.Spec.DatastoreType = apiconfig.Kubernetes
+	if calicoClient, err = client.New(*cfg); err != nil {
+		msg = "failed to initialize api client"
+		return
+	}
+
+	// k8sClient directly calls the k8s apiserver.
+	type accessor interface {
+		Backend() bapi.Client
+	}
+	k8sClient := calicoClient.(accessor).Backend().(*k8s.KubeClient).ClientSet
+
+	// Disable cni
+	if clusterInfo, err = calicoClient.ClusterInformation().Get(ctxt, "default", options.GetOptions{}); err != nil {
+		msg = "failed to fetch cluster information"
+		return
+	}
+	if *clusterInfo.Spec.DatastoreReady {
+		*clusterInfo.Spec.DatastoreReady = false
+		if clusterInfo, err = calicoClient.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
+			msg = "failed to disable cluster"
+			return
+		}
+	}
+
+	// Establishing a mapping of IP addresses to Pods on this node
+	if podList, err = k8sClient.CoreV1().Pods("").List(v1.ListOptions{
+		LabelSelector: fmt.Sprintf("spec.nodeName = %s", nodename),
+	}); err != nil {
+		msg = "failed to list pods"
+		return
+	}
+	podIPMap := make(map[string]*corev1.Pod)
+	for i := 0; i < len(podList.Items); i++ {
+		podIPMap[podList.Items[i].Status.PodIP] = &podList.Items[i]
```

Comment:
We should handle the case where the pod IP is not present.

Reply:
Well this section of the code is just building up the map of IPs to Pods. Handling non-existing pod IPs is done on L171.